src/typechecker.erl

```diff
@@ -1115,33 +1138,77 @@ expect_fun_type_union(Env, [Ty|Tys]) ->
             [TyOut | expect_fun_type_union(Env, Tys)]
     end.
 
-expect_record_type(Record, {type, _, record, [{atom, _, Name}]}, _TEnv) ->
-    if Record == Name ->
-         {ok, constraints:empty()};
-       true ->
-         type_error
+-spec expect_record_type(type(), atom(), #tenv{}) -> Any | FieldsTy | FieldsTys | TypeError when
+    Any :: any,
+    FieldsTy :: {fields_ty, [typed_record_field()], constraints:constraints()},
+    FieldsTys :: {fields_tys, [[typed_record_field()]], constraints:constraints()},
+    TypeError :: {type_error, type()}.
+expect_record_type({user_type, _, record, []}, _Record, _TEnv) ->
+    any;
+expect_record_type({type, _, record, [{atom, _, Name}|RefinedTypes]}, Record, #tenv{records = REnv}) when Name =:= Record ->
+    case REnv of
+        #{Record := Fields} ->
+            Tys =
+                case RefinedTypes of
+                    [] -> Fields;
+                    _ ->
+                        [ {typed_record_field, RecordField, RefinedType}
+                        || {{typed_record_field, RecordField, _}, RefinedType} <- lists:zip(Fields, RefinedTypes)]
+                end,
+            {fields_ty, Tys, constraints:empty()};
+        _NotFound ->
+            {type_error, Record}
     end;
-expect_record_type(_Record, {type, _, tuple, any}, _TEnv) ->
-    {ok, constraints:empty()};
-expect_record_type(_Record, ?top(), _TEnv) ->
-    {ok, constraints:empty()};
-expect_record_type(Record, {type, _, union, Tys}, TEnv) ->
-    expect_record_union(Record, Tys, TEnv);
-expect_record_type(Record, {var, _, Var}, _TEnv) ->
-    {ok, constraints:add_var(Var,
-           constraints:upper(Var, type_record(Record)))};
-expect_record_type(_, _, _) ->
-    type_error.
+expect_record_type(?top() = _TermTy, _Record, _TEnv) ->
+    any;
+expect_record_type({ann_type, _, [_, Ty]}, Record, TEnv) ->
+    expect_record_type(Ty, Record, TEnv);
```

Comment:
It shouldn't be possible to match ann_type here. I have tried to make sure they are removed as early as possible. (Maybe not inside record fields though?)

Reply:
I just copied this code from the `expect_tuple_type` before it was removed I think. I'll make sure it can't happen and remove it.

Reply:
Fixed 👍